columnar/src/columnar/merge/mod.rs

```diff
@@ -348,56 +356,115 @@ fn is_empty_after_merge(
 
-    for &(ref column_name, column_type) in required_columns {
-        columns_grouped
-            .entry((column_name.clone(), column_type.into()))
-            .or_insert_with(|| {
-                GroupedColumns::for_category(column_type.into(), columnar_readers.len())
-            })
-            .require_type(column_type)?;
-    }
+/// Iterates over the columns of the columnar readers, grouped by column name.
+/// Key functionality is that `open` of the Columns is done lazy per group.
+fn group_columns_for_merge_iter<'a>(
+    columnar_readers: &'a [&'a ColumnarReader],
+    required_columns: &'a [(String, ColumnType)],
+    merge_row_order: &'a MergeRowOrder,
+) -> io::Result<impl Iterator<Item = io::Result<(Rc<String>, ColumnType, GroupedColumns)>> + 'a> {
+    let column_iters: Vec<_> = columnar_readers
+        .iter()
+        .enumerate()
+        .map(|(reader_ord, reader)| {
+            Ok(reader
+                .iter_columns()?
+                .map(move |el| (Rc::new(el.0), reader_ord, el.1)))
+        })
+        .collect::<io::Result<_>>()?;
+    let required_columns_map: HashMap<String, _> = required_columns
+        .iter()
+        .map(|(col_name, typ)| (col_name.to_string(), typ))
+        .collect::<HashMap<String, _>>();
+    let mut required_columns_list: Vec<String> = required_columns
+        .iter()
+        .map(|(col_name, _)| col_name.to_string())
+        .collect();
+    required_columns_list.sort();
 
-    for (columnar_id, columnar_reader) in columnar_readers.iter().enumerate() {
-        let column_name_and_handle = columnar_reader.list_columns()?;
-        // We skip columns that end up with 0 documents.
-        // That way, we make sure they don't end up influencing the merge type or
-        // creating empty columns.
+    // Kmerge and group on the column_name.
+    let group_iter = GroupByIteratorExtended::group_by(
+        column_iters.into_iter().kmerge_by(|a, b| a.0 < b.0),
+        |el| el.0.clone(),
+    );
 
-        for (column_name, handle) in column_name_and_handle {
-            let column_category: ColumnTypeCategory = handle.column_type().into();
-            let column = handle.open()?;
-            if is_empty_after_merge(merge_row_order, &column, columnar_id) {
-                continue;
-            }
-            columns_grouped
-                .entry((column_name, column_category))
-                .or_insert_with(|| {
-                    GroupedColumns::for_category(column_category, columnar_readers.len())
-                })
-                .set_column(columnar_id, column);
-        }
-    }
+    // Weave in the required columns into the sorted by column name iterator.
+    let groups_with_required = required_columns_list
+        .into_iter()
+        .merge_join_by(group_iter, |a, b| a.cmp(&b.0));
 
-    let mut merge_columns: BTreeMap<(String, ColumnType), Vec<Option<DynamicColumn>>> =
-        Default::default();
+    Ok(groups_with_required.flat_map(move |either| {
```

Comment:
I wonder if this large and nested iterator construction can be untangled somewhat, at least visually.

IIUC, there is first the construction of `force_type`, `key` and `group`. And there is the the `group_iter.map(..).filter(..)` step based on that.

Maybe these could be separate named functions so that the composition is easier to see? Or at least named let bindings? Admittedly, pulling things apart can actually decrease readability, but it has a bit of a "wall of text" vibe as it is.

Reply:
I refactored it to be more readable